core/src/main/scala/org/apache/spark/executor/ProcfsBasedSystems.scala

```diff
@@ -0,0 +1,226 @@
+        queue ++= c
+        ptree += (p -> c.toSet)
+      }
+      else {
+        ptree += (p -> Set[Int]())
+      }
+    }
+  }
+
+  private def getChildPids(pid: Int): ArrayBuffer[Int] = {
+    try {
+      val cmd = Array("pgrep", "-P", pid.toString)
+      val builder = new ProcessBuilder("pgrep", "-P", pid.toString)
+      val process = builder.start()
+      val output = new StringBuilder()
+      val threadName = "read stdout for " + "pgrep"
+      def appendToOutput(s: String): Unit = output.append(s).append("\n")
+      val stdoutThread = Utils.processStreamByLine(threadName,
+        process.getInputStream, appendToOutput)
+      val exitCode = process.waitFor()
+      stdoutThread.join()
+      // pgrep will have exit code of 1 if there are more than one child process
+      // and it will have a exit code of 2 if there is no child process
+      if (exitCode != 0 && exitCode > 2) {
+        logError(s"Process $cmd exited with code $exitCode: $output")
+        throw new SparkException(s"Process $cmd exited with code $exitCode")
+      }
+      val childPids = output.toString.split("\n")
+      val childPidsInInt = mutable.ArrayBuffer.empty[Int]
+      for (p <- childPids) {
+        if (p != "") {
+          logDebug("Found a child pid: " + p)
+          childPidsInInt += Integer.parseInt(p)
+        }
+      }
+      childPidsInInt
+    } catch {
+      case e: IOException => logDebug("IO Exception when trying to compute process tree." +
+        " As a result reporting of ProcessTree metrics is stopped", e)
+        isAvailable = false
+        return mutable.ArrayBuffer.empty[Int]
+    }
+  }
+
+  def computeProcessInfo(pid: Int): Unit = {
+    /*
+   * Hadoop ProcfsBasedProcessTree class used regex and pattern matching to retrive the memory
+   * info. I tried that but found it not correct during tests, so I used normal string analysis
+   * instead. The computation of RSS and Vmem are based on proc(5):
+   * http://man7.org/linux/man-pages/man5/proc.5.html
+   */
+    try {
+      val pidDir = new File(procfsDir, pid.toString)
+      Utils.tryWithResource( new InputStreamReader(
+        new FileInputStream(
+          new File(pidDir, procfsStatFile)), Charset.forName("UTF-8"))) { fReader =>
+        Utils.tryWithResource( new BufferedReader(fReader)) { in =>
+          val procInfo = in.readLine
+          val procInfoSplit = procInfo.split(" ")
+          if (procInfoSplit != null) {
+            val vmem = procInfoSplit(22).toLong
+            val rssPages = procInfoSplit(23).toLong
+            if (procInfoSplit(1).toLowerCase(Locale.US).contains("java")) {
+              allMetrics = ProcfsBasedSystemsMetrics(
+                allMetrics.jvmVmemTotal + vmem,
+                allMetrics.jvmRSSTotal + (rssPages*pageSize),
+                allMetrics.pythonVmemTotal,
+                allMetrics.pythonRSSTotal,
+                allMetrics.otherVmemTotal,
+                allMetrics.otherRSSTotal
+              )
```

Comment:
case classes have a `copy` method which helps with this

```scala
val rssMem = procInfoSplit(23).toLong * pageSize
...
allMetrics = allMetrics.copy(
  jvmVmemTotal = allMetrics.jvmVMemTotal + vmem, 
  jvmRSSTotal = allMetrics.rssTotal + rssMem
)
```


(also maybe easier if values were written directly into result array, as I mention below)